pkg/diagnostics/diagnostics.go

```diff
@@ -0,0 +1,121 @@
+package diagnostics
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/threefoldtech/zbus"
+	"github.com/threefoldtech/zos/pkg/perf"
+	"github.com/threefoldtech/zos/pkg/utils"
+)
+
+const (
+	callTimeout    = 3 * time.Second
+	testNetworkKey = "perf.network-health"
+)
+
+// Modules is all the registered modules on zbus
+var Modules = []string{
+	"storage",
+	"node",
+	"identityd",
+	"vmd",
+	"flist",
+	"network",
+	"container",
+	"provision",
+	"gateway",
+	"qsfsd",
+}
+
+type moduleStatus struct {
+	Status zbus.Status `json:"status,omitempty"`
+	Err    error       `json:"error,omitempty"`
+}
+
+// Diagnostics show the health of zbus modules
+type Diagnostics struct {
+	// SystemStatusOk is the overall system status
+	SystemStatusOk bool `json:"system_status_ok"`
+	// ZosModules is a list of modules with their objects and workers
+	ZosModules map[string]moduleStatus `json:"modules"`
+	// Online is the state of the grid services reachable from the node
+	Online bool `json:"online"`
+}
+
+func GetSystemDiagnostics(ctx context.Context, msgBrokerCon string) (Diagnostics, error) {
+	busClient, err := zbus.NewRedisClient(msgBrokerCon)
+	if err != nil {
+		return Diagnostics{}, errors.Wrap(err, "fail to connect to message broker server")
+	}
+
+	results := Diagnostics{
+		SystemStatusOk: true,
+		ZosModules:     make(map[string]moduleStatus),
+	}
+
+	var wg sync.WaitGroup
+	for _, module := range Modules {
+
+		wg.Add(1)
+		go func(module string) {
+			defer wg.Done()
+			report := getModuleStatus(ctx, busClient, module)
+			results.ZosModules[module] = report
+
+			if report.Err != nil {
+				results.SystemStatusOk = false
+			}
+		}(module)
+
+	}
+
+	wg.Wait()
+
+	results.Online = isOnline(ctx, msgBrokerCon)
+
+	return results, nil
+
+}
+
+func getModuleStatus(ctx context.Context, busClient zbus.Client, module string) moduleStatus {
+	ctx, cancel := context.WithTimeout(ctx, callTimeout)
+	defer cancel()
+
+	status, err := busClient.Status(ctx, module)
+	return moduleStatus{
+		Status: status,
+		Err:    err,
+	}
+}
+
+func isOnline(ctx context.Context, msgBrokerCon string) bool {
+	client, err := utils.NewRedisPool(msgBrokerCon)
+	if err != nil {
+		return false
+	}
+
+	conn := client.Get()
+	defer conn.Close()
```

Comment:
you closing the connection but not the pool. The idea of a `Pool` is the reusability so creating a pool ONLY to use one connection then through it away does not sound like a good way to use a pool

Instead Pools are created to live long, and used repeatedly to retrieve (and reuse) connections. So at least here you need to close the pool after you re done .